### PR TITLE
fix: treat empty notebook summary as "" not safe_index drift (#1485)

### LINKED ISSUE #1485: Notebook summary fails with safe_index drift on a summary-less (empty/new) notebook
## Summary

Getting a notebook's summary fails with a `safe_index drift` RPC error when the notebook has **no summary yet** (e.g. a brand-new notebook with no sources). A legitimately-empty summary state is being mis-classified as wire-schema drift and raised, instead of being treated as "no description available."

## Symptom

```
RPC: safe_index drift at path ()[0] [method_id='VfAZjd', path=(), source='_notebooks._extract_summary', data_at_failure='None'] (retriable=false)
```

## Affected surfaces (shared layer)

All three go through `notebooklm._notebooks._extract_summary`:
- **CLI**: `notebooklm summary <empty-notebook>` (`src/notebooklm/cli/notebook_cmd.py:267`)
- **Python API**: `client.notebooks.get_summary(notebook_id)` (`src/notebooklm/_notebooks.py:572`)
- **MCP**: the `notebook_describe` tool

## Root cause

`_extract_summary` (`src/notebooklm/_notebooks.py:39-61`) *intends* to return `""` for a missing summary — its docstring says so, and it has an `if summary_val is None: return ""` guard. But that guard only catches a missing **inner** slot. For a summary-less notebook the SUMMARIZE `result[0]` payload (`outer`) is `None`, so the **outer** `[0]` descent inside

```python
summary_val = safe_index(
    outer, 0, 0,
    method_id=RPCMethod.SUMMARIZE.value,
    source="_notebooks._extract_summary",
)
```

treats indexing `None` at the root as schema **drift** and raises before the `None` guard is ever reached (`data_at_failure='None'` at `path=()[0]`).

An empty/absent summary is a routine, legitimate state — not a wire-shape change — so it should not surface as `safe_index drift`.

## Reproduction

```python
async with NotebookLMClient.from_storage() as client:
    nb = await client.notebooks.create("empty")     # no sources
    await client.notebooks.get_summary(nb.id)        # -> RPC safe_index drift
```
or via the CLI: create a fresh notebook, then `notebooklm summary <id>`.

Confirmed the inverse: `get_summary` / `notebook_describe` on **populated** notebooks returns the real AI summary correctly — so this is specific to the no-summary case, not broad drift.

## Proposed fix

Treat the outer `[0]` slot as routinely-optional (return `""` when `outer` / `outer[0]` is `None`/absent), mirroring how `_extract_suggested_topics` already treats its optional `[1]` slot (`_notebooks.py:71-77`) — "missing-slot is not drift." Reserve `safe_index` drift for a present-but-malformed payload. Add a regression test for the empty-notebook path.

## Discovered by

The live MCP-server e2e suite (`tests/e2e/test_mcp.py`), whose `create → describe → …` lifecycle describes a freshly-created (empty) notebook. The MCP layer behaved correctly — it faithfully surfaced the lower-layer failure as a structured `RPC (retriable=false)` error. This is a pre-existing shared-layer bug, not MCP-specific.

## Related

Connects to the positional-RPC-decoding convergence (#1377).

### PR DESCRIPTION
**Pull request by teng-lin**
## What & why

Fixes #1485. A brand-new, source-less notebook has no summary yet, so the `SUMMARIZE` RPC returns an absent/`None` `result[0]` payload. The extraction helpers descended straight into that slot with `safe_index`, so the routine "no summary yet" state raised `UnknownRPCMethodError` ("safe_index drift") — breaking `client.notebooks.get_summary()`, `get_description()`, the CLI `summary` command, and the MCP `notebook_describe` path for **any** empty notebook.

## Approach

Distinguish genuine **absence** from genuine **drift**, and unify the two entry points:

- **`_extract_summary`** — short-circuit to `""` only for `outer is None`, an empty `outer` list, or `outer[0] is None` (the routine "no summary yet" shapes). Everything else descends via `safe_index(outer, 0, 0)`: a scalar `outer`, or a non-`None` `outer[0]` that isn't the expected `[summary_string, ...]` list, is present-but-malformed and still **raises** drift. (Earlier this over-suppressed a scalar/string `outer` to `""`.)
- **`get_summary`** now delegates to `_extract_summary(result[0])`, so it and `get_description` resolve `result[0][0][0]` identically — **single source of truth**. Both agree on every shape: `None`/`[]`/`[None]`/`[[]]`/`[[None]]` → `""`; malformed/scalar/string → raise.
- **`safe_index`** rejects a `str`/`bytes`/`bytearray` value at an **intermediate** descent hop. A string is indexable but never a valid container in a decoded RPC payload, so descending it (`"abc"[0] → "a"`) would smuggle a single character past drift detection. A string **leaf** (the returned final value) is unaffected. This hardens the shared primitive for all 45 call sites (aligns with #1377).

## Tests

| Shape | `get_summary` / `get_description` |
|---|---|
| `None`, `[]`, `[None]`, `[[]]`, `[[None]]` (absence) | `""` |
| `[["the summary"]]` (happy) | `"the summary"` |
| `[[]]`-inner, `[42]`, scalar `123`, `"abc"`, `["summary"]` (malformed) | **raise** `UnknownRPCMethodError` |

- `tests/unit/test_safe_index.py` — str/bytes intermediate hops raise; str leaf returned untouched.
- `tests/unit/test_notebooks_extractors.py`, `tests/unit/test_get_summary_drift.py`, `tests/unit/test_swallow_observability.py`, `tests/integration/test_notebooks_integration.py` — full absence/drift/consistency matrix.

## Verification

- Full suite green (8550 passed; the only failure is the pre-existing, gitignored `.sisyphus/` ADR-format lint that CI never sees). mypy + ruff clean. All 738 repo_lint gates pass.
- API-compat audit `EXIT=0`, **zero** allowlist additions — not an API-breaking change.
- Polished: claude + codex review, both rounds of codex findings (scalar over-suppression, then the `safe_index` string hole) addressed; codex final verdict **RESOLVED**.

Closes #1485.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Empty or missing notebook summaries now return an empty string instead of causing an error.
  * Detection of malformed summary payloads is stricter (e.g., rejects unexpected string/byte-like structures), so only truly malformed responses raise errors.

* **Tests**
  * Expanded edge-case coverage to validate absent vs. malformed summaries and the stricter payload handling.

### CODE REVIEW

**Comment by teng-lin** [comment]
@claude review

File: src/notebooklm/_notebooks.py [hunk 1]
```diff
@@ -40,15 +40,33 @@ def _extract_summary(outer: Any) -> str:
...
-        summary as "no description available").
+        The summary string, or ``""`` when the payload omits the summary
+        slot (the caller is responsible for treating an empty summary as
+        "no description available").
     """
+    # Genuinely-absent summary (no payload, empty payload, or null slot) is the
+    # routine "no summary yet" case — return "" without logging drift.
+    if outer is None:
+        return ""
+    if isinstance(outer, list) and (len(outer) < 1 or outer[0] is None):
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_notebooks.py]: Addressed in eab21e32: switched to implicit falsiness (`not outer` / `not result`) per the repo styleguide. Behavior is identical — the value is `True` exactly when the list is empty, and the `or` short-circuit keeps the `outer[0]` access guarded. Thanks!

File: src/notebooklm/_notebooks.py [hunk 2]
```diff
@@ -586,16 +604,13 @@ async def get_summary(self, notebook_id: str) -> str:
...
-            0,
-            method_id=RPCMethod.SUMMARIZE.value,
-            source="_notebooks.get_summary",
-        )
-        return str(summary) if summary else ""
+        # Response structure: [[[summary_string, ...], topics, ...]]. ``result[0]``
+        # is the ``outer`` payload that ``_extract_summary`` descends, so delegate
+        # to it: empty/None/null-slot → "" and present-but-malformed → drift,
+        # identically to ``get_description`` (single source of truth — #1485).
+        if not isinstance(result, list) or len(result) < 1:
```
[the remark this answers is not in the document]
Comment (teng-lin) [reply src/notebooklm/_notebooks.py]: Addressed in eab21e32: switched to implicit falsiness (`not outer` / `not result`) per the repo styleguide. Behavior is identical — the value is `True` exactly when the list is empty, and the `or` short-circuit keeps the `outer[0]` access guarded. Thanks!

**Comment by teng-lin** [comment]
Thanks @claude. Addressed the one actionable nit in 2da909d8: corrected the comment in `test_get_summary_drift_raises_typed_error` — the payload is `[[[]]]`, so `result[0]` (== `outer`) is `[[]]` (present, non-None) and it's the summary slot `result[0][0]` that is the empty list driving the drift at `result[0][0][0]`. The guard-polarity difference between `get_summary` and `get_description` is cosmetic (both delegate to `_extract_summary` as the single source of truth), so I've left it as-is.
